### ISSUE TITLE
Improved MySql Searching for exact matches over partial matches

### ISSUE THREAD

**Issue by AirmanAJK** [magento/magento2#13866]
The built-in MySql search in Magento does not favor exact matches of search terms over matches that simply contain a search term.

### Preconditions

Magento 2.4-develop

1. Any Magento 2 version using the MySQL search using name as an indexed search attribute

### Steps to reproduce
Set up several product with names such as:
Prod
Product
Products

### Expected result
Searching just "Prod" should favor the 1st product in the above list over the 2nd and 3rd products. It is an exact match, while the others are partial matches.

### Actual result
Magento's implementation of search ranks them equally, and the less relevant results may be returned as a higher priority.

To fix:
In the file "Magento\Framework\Search\Adapter\Mysql\Query\Builder\Match" there is a function "prepareQuery"

This is the line of interest (141 in the latest version):
$queryValues[$queryKey] = $stringPrefix . $queryValue . $stringSuffix;

I believe this can be greatly improved by also adding this string to the queryValues collection:
`'(>' . $queryValue . ') '`

This will still include the same search terms, but give increased relevance to the exact match over the partial matches. We have short model numbers on our web site (such as simply 437) and we definitely wanted the 437 to show up first, and the other models like 437231 and "AB 4371 " to show up after. This fixes the problem.

**Comment by fooman** [comment]
@AirmanAJK thanks a lot for your report. Would you be interested in supplying your suggested changes as a pull request?

**Comment by magento-engcom-team** [comment]
@AirmanAJK, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by ghost** [comment]
Hi Guys, Is this going to be released in Magento v2.3 ? 

thanks,

**Comment by stephansteiner** [comment]
Still an issue in 2.2.5 - still no patch available?

**Comment by AirmanAJK** [comment]
I wouldn't expect anything anytime soon. I used di.xml to override the Match.php file to replace the loop near the bottom of the prepareQuery function with the following:

```
foreach ($queryValues as $queryKey => $queryValue) {
	if (empty($queryValue) || strlen($queryValue) == 1) {
		unset($queryValues[$queryKey]);
	} else if($stringPrefix == '') {
		$queryValues[$queryKey] = '>' . $queryValue . ' <' . $queryValue . '*';
	} else {
		$queryValues[$queryKey] = $stringPrefix . $queryValue . '*';
	}
}
```

The elseif statement will give increased relevance to exact matches, while decreasing relevance for partial matches. As always, test it like crazy if put into production.

**Comment by Ctucker9233** [comment]
@magento-engcom-team What version of 2.2 or 2.3 can we expect this in?

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Echo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-30508` were created

**Issue Available**: @engcom-Echo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by engcom-Bravo** [comment]
Hello @AirmanAJK 

Thank you for your report

Seems like the issue is not reproducible on the latest Magento 2.4-develop
![prod_search](https://user-images.githubusercontent.com/51680745/94054870-de2bba80-fde4-11ea-8943-d11214fd58bb.png)

The latest Magento version uses Elasticsearch engine instead of MySQL.

Can you please check if you can still reproduce the issue on the latest 2.4-develop? If it is not reproducible, we will have to close this issue.

Thank you for collaboration

**Comment by sdzhepa** [comment]
Closed as cannot reproduce on the latest `2.4-develop` code.
But if you still faced the issue please create a new report for 2.4 with detailed steps
